src/RealFourierArray.jl

```diff
@@ -0,0 +1,91 @@
+## TODO implement compressed representation in Fourier space of real arrays
+
+# Tr is the underlying real type
+# T is the type of the array in real space
+# TODO make this concrete by parametrizing on the type of the arrays
+mutable struct RealFourierArray{Tr <: Real, T <: Union{Tr, Complex{Tr}}}
+    basis::PlaneWaveBasis{Tr}
+    _real::AbstractArray{T, 3}
+    _fourier::AbstractArray{Complex{Tr}, 3}
+    _state::Int # 0: only _real set. 1: only _fourier set. 2: both set
+end
+# Type of the real part
+Base.eltype(A::RealFourierArray{Tr, T}) where {Tr, T} = T
+
+# Constructors
+RealFourierArray(basis, real, fourier) = RealFourierArray(basis, real, fourier, 2)
+function RealFourierArray(basis::PlaneWaveBasis{T}; iscomplex=false) where {T}
+    from_real(basis, zeros(iscomplex ? complex(T) : T, basis.fft_size...))
+end
+
+function from_real(basis, real_part::AbstractArray{T}) where {T <: Number}
+    RealFourierArray{real(T), T}(basis, real_part, similar(real_part, complex(T)), 0)
+end
+# this will necessarily be a real array
+function from_fourier(basis, fourier_part::AbstractArray{T}; isreal=false) where {T <: Complex}
+    if isreal
+        RealFourierArray{real(T), real(T)}(basis, similar(fourier_part, real(T)), fourier_part, 1)
+    else
+        RealFourierArray{real(T), T}(basis, similar(fourier_part), fourier_part, 1)
+    end
+end
+
+check_real(A::RealFourierArray{Tr, T}) where {Tr, T <: Real} = nothing
+check_real(A::RealFourierArray) = check_real(A.real)
+
+# lie blatantly to the user, as recommended by the docs
+Base.fieldnames(RealFourierArray) = (:basis, :real, :fourier)
+# TODO optimize this
+function Base.getproperty(A::RealFourierArray, x::Symbol)
+    if x == :real
+        if A._state == 1
+            r = G_to_r(A.basis, A._fourier)
+            if eltype(A) <: Real
+                A._real .= real(r)
+            else
+                A._real .= r
+            end
+            setfield!(A, :_state, 2)
+        end
+        return A._real
+    elseif x == :fourier
+        if A._state == 0
+            r_to_G!(A._fourier, A.basis, complex.(A._real))
+            setfield!(A, :_state, 2)
+        end
+        return A._fourier
+    else
+        getfield(A, x)
+    end
+end
+
+function Base.setproperty!(A::RealFourierArray, x)
+    error("RealFourierArray is intended to be read-only.
+(This can be bypassed by `setfield!` if you really want to).")
+end
+
+# TODO optimize those a bit better
+function Base.:+(A::RealFourierArray, B::RealFourierArray) where T
+    @assert A.basis == B.basis
+    from_real(A.basis, A.real + B.real)
+end
+function Base.:-(A::RealFourierArray, B::RealFourierArray) where T
+    @assert A.basis == B.basis
+    from_real(A.basis, A.real - B.real)
+end
+function Base.:*(α::Number, A::RealFourierArray)
+    from_real(A.basis, α*A.real)
+end
+Base.:*(A::RealFourierArray, α::Number) = α*A
+Base.:/(A::RealFourierArray, α::Number) = A*inv(α)
```

Comment:
Spaces around operators ;)

Reply:
Argh. I think my instincts push me towards no space for multiplication, to make precedence clearer. Ie I would naturally write `a*x + b`. julia source seems to be pretty lax with this. Eg looking at a random file (dense.jl in LinearAlgebra), I see
```
        U  = A * (A6 * (CC[14].*A6 .+ CC[12].*A4 .+ CC[10].*A2) .+
                  CC[8].*A6 .+ CC[6].*A4 .+ CC[4].*A2 .+ CC[2].*Inn)
```
where I think the spacing improves readability.

Reply:
I see your point, but my experience is that actually doing the same spacing around all operators helps. E.g. python enforces this and I made good experiences with it. But I get that this could be a bit of personal preference. Anyway ... we wanted to look into `JuliaFormatter` at some point anyway, so I suggest we use that to fix formatting. I think they actually use the spaces, however ;)